cherrypy/test/test_client_cert.py

```diff
@@ -0,0 +1,118 @@
+"""Tests for client certificate validation."""
+
+import itertools
+import ssl
+from os.path import abspath, basename, dirname, join
+from six.moves import urllib
+
+import cherrypy
+from cherrypy.test import helper
+
+
+SSL_DIR = join(abspath(dirname(__file__)), 'ssl')
+CA_CERT = join(SSL_DIR, 'ca.cert')
+CLIENT_KEY = join(SSL_DIR, 'client.key')
+CLIENT_CERT = join(SSL_DIR, 'client.cert')
+CLIENT_IP_CERT = join(SSL_DIR, 'client_ip.cert')
+CLIENT_WILD_CERT = join(SSL_DIR, 'client_wildcard.cert')
+CLIENT_WRONG_CA = join(SSL_DIR, 'client_wrong_ca.cert')
+CLIENT_WRONG_HOST = join(SSL_DIR, 'client_wrong_host.cert')
+SERVER_KEY = join(SSL_DIR, 'server.key')
+SERVER_CERT = join(SSL_DIR, 'server.cert')
+
+
+class ClientCertTests(object):
+    """Client certificate validation tests."""
+    scheme = 'https'
+    script_name = '/index'
+
+    server_should_reject = False
+    server_host = 'localhost'
+    client_host = 'localhost'
+    ca_chain = CA_CERT
+    client_cert = CLIENT_CERT
+    verify_mode = 'none'  # none, optional, required
+
+    @classmethod
+    def setup_server(cls):
+        cherrypy.config.update({
+            'checker.on': False,
+            'engine.autoreload_on': False,
+            'server.socket_host': cls.server_host,
+            'server.ssl_module': 'builtin',
+            'server.ssl_private_key': SERVER_KEY,
+            'server.ssl_certificate': SERVER_CERT,
+            'server.ssl_certificate_chain': cls.ca_chain,
+            'server.ssl_verify_mode': cls.verify_mode,
+        })
+
+        class Root:
+            @cherrypy.expose
+            def index(self):
+                return 'ok'
+
+        cherrypy.tree.mount(Root())
+
+    def test_connect(self):
+        """If server_should_reject is false, verify the index can be fetched.
+        Otherwise, verify that a URLError is raised when connecting.
+        """
+        context = ssl.create_default_context(cafile=self.ca_chain)
+        context.load_cert_chain(self.client_cert, keyfile=CLIENT_KEY)
+        context.check_hostname = False
+
+        if self.server_should_reject:
+            self.assertRaises(
```

Comment:
We are trying to move away from pure `unittest`. Could you please re-write this using pytest?